File: lib/syntax/body_instruction.cpp
/*
 * body_instruction.cpp
 *
 *  Created on: Jul 24, 2020
 *      Author: Matuzalem Muller
 */

#include "body_instruction.h"

BodyInstruction::BodyInstruction(BodyType type,
                                 Statement stm,
                                 bool (*take_action)())
{
  _type = type;
  _statement = stm;
  _take_action = take_action;
  _operator = EventOperator::BELIEF_ADDITION;
}

BodyInstruction::BodyInstruction(BodyType type,
                                 Statement stm,
                                 EventOperator event_operator)
{
  _type = type;
  _statement = stm;
  _take_action = nullptr;
  _operator = event_operator;
}

// Add handling of belief that is not in beliefbase
BodyReturn BodyInstruction::run_instruction(BeliefBase * belief_base,
                                            EventBase * event_base)
{
  BodyReturn result;
  Event event(_operator, _statement);

  if (_type == BodyType::ACTION)
  {
    bool value = _take_action();
    result = BodyReturn(BodyType::ACTION, value, nullptr);
  }
  else if (_type == BodyType::BELIEF)
  {
    if (event_base->add_event(event))
    {
      if (_operator == EventOperator::BELIEF_ADDITION)
      {
        belief_base->change_belief_state(_statement, true);
      }
      else
      {
        belief_base->change_belief_state(_statement, false);
      }
      result = BodyReturn(BodyType::BELIEF,
                          true,
                          event_base->last_event()->get_event_id()
                          );
    }
    else
    {
      result = BodyReturn(BodyType::BELIEF, false, nullptr);
    }
  }
  else
  {
    if (event.get_operator() == EventOperator::GOAL_ACHIEVE)
    {
      event_base->add_event(event);
      result = BodyReturn(
                          BodyType::GOAL,
                          true,
                          nullptr
                          );
    }
    else if (event_base->add_event(event))
    {
      result = BodyReturn(
                          BodyType::GOAL,
                          true,
                          event_base->last_event()->get_event_id()
                          );
    }
    else
    {
      result = BodyReturn(BodyType::GOAL, false, nullptr);
    }
  }

  return result;
}

File: config/java/src/translator/BodyInstruction.java
package translator;

public class BodyInstruction {
  enum BodyType {
    ACTION,
    BELIEF,
    GOAL
  };

  private BodyType type;
  private String statement;
  private EventOperatorType operator;

  public BodyInstruction() {
  }

  public BodyInstruction(BodyType type, String statement)
  {
    this.type = type;
    this.statement = statement;
  }

  public BodyInstruction(BodyType type, String statement, EventOperatorType operator)
  {
    this.type = type;
    this.statement = statement;
    this.operator = operator;
  }

  public BodyType getType() {
    return type;
  }

  public void setType(BodyType type) {
    this.type = type;
  }

  public String getStatement() {
    return statement;
  }

  public void setStatement(String statement) {
    this.statement = statement;
  }

  public EventOperatorType getOperator() {
    return operator;
  }

  public void setOperator(EventOperatorType operator) {
    this.operator = operator;
  }
}